File: apps/ios/ios/NativeToReact/ContextMenuManager/ContextMenuManager.m
//
//  ContextMenuManager.m
//  ios
//
//  Created by Jignesh Raiyani on 10/6/16.
//  Copyright © 2016 QuestionPro Pvt Ltd. All rights reserved.
//

#import "ContextMenuManager.h"
#import <UIKit/UIKit.h>

@implementation ContextMenuManager

+ (id)allocWithZone:(NSZone *) zone
{
    static ContextMenuManager *sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [super allocWithZone:zone];
    });
    return sharedInstance;
}

RCT_EXPORT_MODULE();

- (NSArray<NSString*> *)supportedEvents {
    return @[@"ContextMenuItemClick",@"SceneTransition", @"reloadScreen", @"Logout", @"GoalsFilterAction", @"ObjEditAction", @"LanguagePicker"];
}

#pragma mark - send event from ObjcetiveC to React

-(void)refreshContextMenuWithdata:(NSDictionary*)aContextMenuDict {
    
    NSError * error;
    NSData * jsonData = [NSJSONSerialization dataWithJSONObject:aContextMenuDict options:0 error:&error];
    NSString * jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    NSDictionary *dataDict = @{@"DATA":jsonString};
    [self sendEventWithName:@"ContextMenuItemClick" body:dataDict];
    
}

#pragma mark - send event from ObjcetiveC to React

-(void)refreshProfileScreenWithdata:(NSString*)aProfileMenuDict withTitle:(NSString*)aTitle {

    NSDictionary *dataDict = @{@"Scene":aProfileMenuDict, @"Title": aTitle};
    [self sendEventWithName:@"SceneTransition" body:dataDict];
}

-(void)reloadHomeScreenForPulse {
    [self sendEventWithName:@"reloadScreen" body:@""];
}

-(void)showLanguagePicker {
    [self sendEventWithName:@"LanguagePicker" body:@""];
}

-(void)logoutUser {
    [self sendEventWithName:@"Logout" body:@""];
}

//@sujan Change objectives and goals screen based on filter options
-(void)reloadObjAndGoalsScreen:(NSString*)option {
    
    NSDictionary *dataDict = @{@"filter":option};
    [self sendEventWithName:@"GoalsFilterAction" body:dataDict];
}

//@sujan Edit the selected objectives and goals
-(void)editObjAndGoal {
    [self sendEventWithName:@"ObjEditAction" body:@""];
}

@end

File: apps/ios/ios/NativeToReact/ContextMenuManager/ContextMenuManager.swift
//
//  ContextMenuManager.swift
//  ios
//
//  Created by Jignesh on 14/05/18.
//  Copyright © 2018 Jignesh. All rights reserved.
//

import Foundation

@objc(ContextMenuManager)
class ContextMenuManager:  RCTEventEmitter {
    
    @objc override func supportedEvents() -> [String]! {
        return ["ContextMenuItemClick", "SceneTransition", "Logout", "reloadScreen", "GoalsFilterAction", "ObjEditAction", "LanguagePicker"];
    }
    
    @objc(refreshProfileScreenWithdata:aTitle:)
    func refreshProfileScreenWithdata(aProfileMenu : NSString, aTitle : NSString) {
        let dataDict : [String:Any] = ["Scene":aProfileMenu, "Title":aTitle]
        self.sendEvent(withName: "SceneTransition", body: dataDict)
    }
    
    @objc func logoutUser() {
        self.sendEvent(withName: "Logout", body: "")
    }
    
    @objc func showLanguagePicker() {
        self.sendEvent(withName: "LanguagePicker", body: "")
    }
    
    @objc func refreshContextMenuWithdata(aContextMenuDict : [String:Any]) {
        var dataDict = [String:Any]()
        if let theJSONData = try? JSONSerialization.data(withJSONObject: aContextMenuDict,options: []) {
            let theJSONText = String(data: theJSONData,encoding: .utf8)
            dataDict = ["DATA":theJSONText ?? ""]
            print("JSON string = \(theJSONText!)")
        }
        self.sendEvent(withName: "ContextMenuItemClick", body: dataDict)
    }
    
    @objc func reloadHomeScreenForPulse() {
        self.sendEvent(withName: "reloadScreen", body: "")
    }
    
    @objc func reloadObjAndGoalsScreen(option : NSString) {
        let dataDict : [String:Any] = ["filter":option]
        self.sendEvent(withName: "GoalsFilterAction", body: dataDict)
    }
    
    @objc func editObjAndGoal() {
        self.sendEvent(withName: "ObjEditAction", body: "")
    }
    
    
}
